src/Plugin/Field/FieldWidget/ExploreCoursesUrlWidget.php

```diff
@@ -0,0 +1,188 @@
+    $elements['api_version'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('API version'),
+      '#required' => TRUE,
+      '#default_value' => $this->getSetting('api_version'),
+      '#access' => TRUE,
+      '#element_validate' => [
+        [$this, 'validateApi'],
+      ],
+    ];
+    return $elements;
+  }
+
+  /**
+   * Validate we are specifying the newest api.
+   *
+   * @param array $element
+   *   Url form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Current form state object.
+   * @param array $complete_form
+   *   Complete form.
+   */
+  public function validateApi(array &$element, FormStateInterface $form_state, array &$complete_form) {
+    $input = NestedArray::getValue($form_state->getValues(), $element['#parents']);
+
+    if ($form_state::hasAnyErrors()) {
+      return;
+    }
+    try {
+      $response = $this->client->request('GET', 'search?view=xml-' . $input, ['base_uri' => 'https://explorecourses.stanford.edu/']);
+      $response = (string) $response->getBody();
+      $xml = new \SimpleXMLElement($response);
+      // Do this as a string, since SimpleXMLElement doesn't cast to bools.
+      if ((string) $xml->deprecated == 'true') {
+        $form_state->setError($element, $this->t("That API version is deprecated. Newest version is: $xml->latestVersion"));
+      }
+    }
+    catch (\Throwable $e) {
+      $form_state->setError($element, $this->t('There was a problem querying the ExploreCourses API.'));
+    }
+  }
+
+  /**
+   * {@inheritDoc}.
+   */
+  public function settingsSummary() {
+    $summary = [];
+    if (empty($this->getSetting('api_version'))) {
+      $summary[] = $this->t('No API version Provided');
+    }
+    else {
+      $summary[] = $this->t('API version: @api_version', ['@api_version' => $this->getSetting('api_version')]);
+    }
+    return $summary;
+  }
+
+  /**
+   * {@inheritDoc}.
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $element = parent::formElement($items, $delta, $element, $form, $form_state);
+
+    $element['#element_validate'] = [
+      [$this, 'validateExploreCourseUrl'],
+    ];
+    $element['uri']['#description'] = $this->t('This must be a valid ExplorerCourses URL. See: @url', ['@url' => 'https://explorecourses.stanford.edu']);
+    $element['title']['#access'] = FALSE;
+    $element['attributes']['#access'] = FALSE;
+
+    return $element;
+  }
+
+  /**
+   * Validate we have a legit url.
+   *
+   * @param array $element
+   *   Url form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Current form state object.
+   * @param array $complete_form
+   *   Complete form.
+   */
+  public function validateExploreCourseUrl(array &$element, FormStateInterface $form_state, array &$complete_form) {
+
+    $url = UrlHelper::parse($element['uri']['#value']);
```

Comment:
Does this grab the submitted value? I thought doing it this way only grabs the previously entered value. I think the correct way is to do something like `$form_state->getValue($element['parents'])`. Something like that.

Reply:
It actually does work, but since I attached the validator to the `['uri']` element per above, we can remove the `['uri']` portion from here.  I'm not sure why you'd try to get the value from the parent of the element, but it's coming in as an array on a per-instance basis for the field, so might as well just access the array rather than making a function call.